servers/services/src/main/java/org/projectnessie/services/authz/AccessChecker.java

```diff
@@ -0,0 +1,117 @@
+ */
+public interface AccessChecker {
+
+  /**
+   * Checks whether the given role/principal is allowed to create a Branch/Tag. It is up to the
+   * implementor to define which exception should being thrown in case of denial.
+   *
+   * @param context The context carrying the principal information.
+   * @param ref The {@link NamedRef} to check
+   */
+  void canCreateReference(AccessContext context, NamedRef ref);
+
+  /**
+   * Checks whether the given role/principal is allowed to assign the given Branch/Tag to a Hash. It
+   * is up to the implementor to define which exception should being thrown in case of denial.
+   *
+   * @param context The context carrying the principal information.
+   * @param ref The {@link NamedRef} to check
+   */
+  void canAssignRefToHash(AccessContext context, NamedRef ref);
+
+  /**
+   * Checks whether the given role/principal is allowed to delete a Branch/Tag. It is up to the
+   * implementor to define which exception should being thrown in case of denial.
+   *
+   * @param context The context carrying the principal information.
+   * @param ref The {@link NamedRef} to check
+   */
+  void canDeleteReference(AccessContext context, NamedRef ref);
+
+  /**
+   * Checks whether the given role/principal is allowed to read entries content for the given
+   * Branch/Tag. It is up to the implementor to define which exception should being thrown in case
+   * of denial.
+   *
+   * @param context The context carrying the principal information.
+   * @param ref The {@link NamedRef} to check
+   */
+  void canReadEntries(AccessContext context, NamedRef ref);
+
+  /**
+   * Checks whether the given role/principal is allowed to list the commit log for the given
+   * Branch/Tag. It is up to the implementor to define which exception should being thrown in case
+   * of denial.
+   *
+   * @param context The context carrying the principal information.
+   * @param ref The {@link NamedRef} to check
+   */
+  void canListCommitLog(AccessContext context, NamedRef ref);
+
+  /**
+   * Checks whether the given role/principal is allowed to commit changes against the given
+   * Branch/Tag. It is up to the implementor to define which exception should being thrown in case
+   * of denial.
+   *
+   * @param context The context carrying the principal information.
+   * @param ref The {@link NamedRef} to check
+   */
+  void canCommitChangeAgainstReference(AccessContext context, NamedRef ref);
+
+  /**
+   * Checks whether the given role/principal is allowed to read an entity value as defined by the
+   * {@link ContentsKey} for the given Branch/Tag. It is up to the implementor to define which
+   * exception should being thrown in case of denial.
+   *
+   * @param context The context carrying the principal information.
+   * @param ref The {@link NamedRef} to check
+   * @param key The {@link ContentsKey} to check
+   */
+  void canReadEntityValue(AccessContext context, NamedRef ref, ContentsKey key);
```

Comment:
we may have to provide the contents object or contents.uuid here and in other places contentskey is used. That complicates how/when this is called though

Reply:
agreed that this makes it a bit more complicated

Reply:
Checking against the value of a `Contents` object should be a separate operation applied on the returned values IMO, because otherwise it requires a additional reads before the access check.

Not sure whether it makes a lot of sense to couple the contents with authz rules.

(Checking for the contents-key is fine IMO.)